scalafix-core/shared/src/main/scala/scalafix/internal/rule/NoFinalize.scala

```diff
@@ -0,0 +1,34 @@
+package scalafix.internal.rule
+
+import scala.meta._
+import scala.meta.contrib._
+
+import metaconfig.{Conf, Configured}
+import scalafix.rule.Rule
+import scalafix.rule.RuleCtx
+import scalafix.lint.LintMessage
+import scalafix.lint.LintCategory
+
+case object NoFinalize extends Rule("NoFinalize") {
+  override def check(ctx: RuleCtx): Seq[LintMessage] = {
+    ctx.tree.collect {
+      case defn: Defn.Def if isFinalized(defn) => error(defn.name.pos)
+    }.toSeq
+  }
+
+  private val errorCategory: LintCategory =
+    LintCategory.error(
+      "finalizer may be never called and have a severe performance penalty")
+
+  private def error(pos: Position): LintMessage = errorCategory.at(pos)
+
+  private def isFinalized(defn: Defn.Def): Boolean = {
```

Comment:
wouldn't a quasi-quote be easier to match against in this case?

Something like

```
ctx.tree.collect {
  case q"..override def finalize(): Unit = $_" => // ...
}
```

Reply:
yes, but how do you match against all the cases?

```
override def finalize(): Unit = $_
override protected def finalize(): Unit = $_
override protected def finalize() = $_
```

Reply:
ah, fair point. I guess you can't do fuzzy matching with the quasi-quotes. 

Maybe something like

`case q"..$mods def finalize(): $tpeopt" if mods.exists(_isEqual(Mod.Override)) && ... =>`

but then it's probably not more readable than the existing code. Disregard my comment :D 👍

Reply:
This implementation still does not handle cases like
```scala
class A {
    override def finalize() = println(1)
  }
```

There is no need to check `override` since it's not possible to define a method named "finalize"
```scala
@ class A {
    def finalize = println(1)
  }
cmd0.sc:2: overriding method finalize in class Object of type ()Unit;
 method finalize needs `override' modifier
  def finalize = println(1)
      ^
Compilation Failed
```

I propose to use Defn.Def instead
```scala
@ q"def finalize = 2" match {
    case Defn.Def(_, name @ q"finalize", _, Nil | Nil :: Nil, _, _) => name
  }
res5: Term.Name = Term.Name("finalize")

@ q"def finalize(x: Int) = 2" match {
    case Defn.Def(_, name @ q"finalize", _, Nil | Nil :: Nil, _, _) => name
  }
scala.MatchError: def finalize(x: Int) = 2 (of class scala.meta.Defn$Def$DefnDefImpl)
  ammonite.$sess.cmd6$.<init>(cmd6.sc:2)
  ammonite.$sess.cmd6$.<clinit>(cmd6.sc)
res4: Term.Name = Term.Name("finalize")
```